crates/bitwarden/src/tool/generators/passphrase.rs

```diff
@@ -0,0 +1,112 @@
+use crate::{error::Result, wordlist::EFF_LONG_WORD_LIST};
+use rand::{seq::SliceRandom, Rng, RngCore};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+/// Passphrase generator request.
+///
+/// The default separator is `-` and default number of words is 3.
+#[derive(Serialize, Deserialize, Debug, JsonSchema, Default)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+#[cfg_attr(feature = "mobile", derive(uniffi::Record))]
+pub struct PassphraseGeneratorRequest {
+    pub num_words: Option<u8>,
+    pub word_separator: Option<String>,
+    pub capitalize: Option<bool>,
+    pub include_number: Option<bool>,
+}
+
+const DEFAULT_PASSPHRASE_NUM_WORDS: u8 = 3;
+const DEFAULT_PASSPHRASE_SEPARATOR: char = ' ';
+
+pub(super) fn passphrase(input: PassphraseGeneratorRequest) -> Result<String> {
+    let num_words = input.num_words.unwrap_or(DEFAULT_PASSPHRASE_NUM_WORDS);
+    let separator = input
+        .word_separator
+        .and_then(|s| s.chars().next())
+        .unwrap_or(DEFAULT_PASSPHRASE_SEPARATOR);
+    let capitalize = input.capitalize.unwrap_or(false);
+    let include_number = input.include_number.unwrap_or(false);
+
+    let mut rand = rand::thread_rng();
+
+    let mut passphrase_words = gen_words(&mut rand, num_words);
+    if include_number {
+        include_number_in_words(&mut rand, &mut passphrase_words);
+    }
+    if capitalize {
+        capitalize_words(&mut passphrase_words);
+    }
+    Ok(passphrase_words.join(&separator.to_string()))
+}
+
+fn gen_words(mut rng: impl RngCore, num_words: u8) -> Vec<String> {
+    (0..num_words)
+        .map(|_| {
+            EFF_LONG_WORD_LIST
+                .choose(&mut rng)
+                .expect("slice is not empty")
+                .to_string()
+        })
+        .collect()
+}
```

Comment:
**suggestion**: Let's add a quick test for this function using https://docs.rs/rand/latest/rand/rngs/mock/struct.StepRng.html.

Reply:
I've looked into it, but it seems that the `choose` function expects the numbers to be uniformly distributed in the range of `0..2^64` to give proper results, so a mock rng like StepRng which just returns an incremental counter ends up with the following results, which don't seem that interesting to test:
```
let rng = StepRng::new(0, 1);
println!("{:?}", gen_words(rng, 8));
// Prints: 
["abacus", "abacus", "abacus", "abacus", "abacus", "abacus", "abacus", "abacus"]
```

I've tried to play with the parameters of StepRng a bit but it doesn't help that much either and I get the same results most of the time.

We might be able to use a different RNG, like `rand_chacha`, and just seed it with a constant initial value? This would return the same sequence of values each time, but they'd be more randomly distributed https://docs.rs/rand_chacha/latest/rand_chacha/struct.ChaCha8Rng.html#method.from_seed

Reply:
I think that sounds good, we can add `rand_chacha` as a test dependency.